crates/bevy_ecs/src/component.rs

```diff
@@ -397,6 +397,30 @@ impl SparseSetIndex for ComponentId {
     }
 }
 
+/// A value which associates a raw component id to a statically-known type
+///
+/// This is somewhat niche, but very useful for accessing dynamic components without requiring raw
+/// [`bevy_ptr::Ptr`]s
+pub struct TypedComponentId<T: Send + Sync + 'static> {
+    pub(crate) id: ComponentId, //todo: pub(crate) fields and have the World construct this, or private fields and construct from an &mut World reference?
+    pub(crate) world_id: WorldId,
+    pub(crate) data: PhantomData<T>,
+}
```

Comment:
This should document that a safety invariant of this type is that the world with id `self.world_id` contains a component with id `self.id` of type `T`.